core/Tracker/RequestProcessor.php

```diff
@@ -0,0 +1,165 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Tracker;
+
+use Piwik\Tracker\Visit\VisitProperties;
+
+/**
+ * Base class for all tracker RequestProcessors. RequestProcessors handle and respond to tracking
+ * requests.
+ *
+ * ## Concept: Request Metadata
+ *
+ * RequestProcessors take a Tracker\Request object and based on its information, set request metadata.
+ *
+ * Request metadata is information about the current tracking request, for example, whether
+ * the request is for an existing visit or new visit, or whether the current visitor is a known
+ * visitor, etc. It is used to control tracking behavior.
+ *
+ * Request metadata is shared between RequestProcessors, so RequestProcessors can tweak each others
+ * behavior, and thus, the behavior of the Tracker. Request metadata can be set and get using the
+ * {@link VisitProperties::setRequestMetadata()} and {@link VisitProperties::getRequestMetadata()}
+ * methods.
+ *
+ * Each RequestProcessor lists the request metadata it computes and exposes in its class
+ * documentation.
+ *
+ * ## The Tracking Process
+ *
+ * When Piwik handles a single tracking request, it gathers all available RequestProcessors and
+ * invokes their methods in sequence.
+ *
+ * The first method called is {@link self::processRequestParams()}. RequestProcessors should use
+ * this method to compute request metadata and set visit properties using the tracking request.
+ * An example includes the ActionRequestProcessor, which uses this method to determine the action
+ * being tracked.
+ *
+ * The second method called is {@link self::afterRequestProcessed()}. RequestProcessors should
+ * use this method to either compute request metadata/visit properties using other plugins'
+ * request metadata, OR override other plugins' request metadata to tweak tracker behavior.
+ * An example of the former can be seen in the GoalsRequestProcessor which uses the action
+ * detected by the ActionsRequestProcessor to see if there are any action-matching goal
+ * conversions. An example of the latter can be seen in the PingRequestProcessor, which on
+ * ping requests, aborts conversion recording and new visit recording.
+ *
+ * After these two methods are called, either {@link self::onNewVisit()} or {@link self::onExistingVisit()}
+ * is called. Generally, plugins should favor defining Dimension classes instead of using these methods,
+ * however sometimes it is not possible (as is the case with the CustomVariables plugin).
```

Comment:
If it's only a special case for custom variables, maybe events would be more appropriate to keep things for normal cases easy and simple (eg less methods and docs to read etc)? Not so much into it so up to you. Just saying events can be beneficial in such cases even though I'm not a fan of them either